// @microsoft/power-apps-cli/dist/HttpClient/CliHttpClient.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { spinner } from '@clack/prompts';
import { randomUUID } from 'crypto';
import { getSettings } from '../CliSettings.js';
import { MS_CLIENT_REQUESTID } from '../Constants.js';
import { enhanceNetworkError } from '../Utils/EnhanceNetworkError.js';
export class CliHttpClient {
    _authenticationProvider;
    _inflight = 0;
    _spinner = spinner();
    constructor(_authenticationProvider) {
        this._authenticationProvider = _authenticationProvider;
    }
    async get(url, config) {
        return this._sendRequestAsync(url, 'GET', config);
    }
    async post(url, config) {
        return this._sendRequestAsync(url, 'POST', config);
    }
    async patch(url, config) {
        return this._sendRequestAsync(url, 'PATCH', config);
    }
    async put(url, config) {
        return this._sendRequestAsync(url, 'PUT', config);
    }
    async _sendRequestAsync(url, method, config) {
        this._startSpinner();
        try {
            const headers = config?.headers || {};
            if (config?.authResource) {
                const token = await this._authenticationProvider.getAccessTokenForResource(config.authResource);
                headers.Authorization = `Bearer ${token}`;
            }
            if (!headers['Content-Type']) {
                headers['Content-Type'] = 'application/json';
            }
            headers[MS_CLIENT_REQUESTID] = randomUUID();
            // Only JSON.stringify if body is not already a string and Content-Type is JSON
            let bodyContent;
            if (config?.body !== undefined) {
                const isJsonContentType = headers['Content-Type'] === 'application/json';
                bodyContent =
                    typeof config.body === 'string' && !isJsonContentType
                        ? config.body
                        : JSON.stringify(config.body);
            }
            const requestOptions = {
                method,
                headers,
                // Only include body if it exists
                body: bodyContent,
            };
            const response = await fetch(url, requestOptions);
            if (!response.ok) {
                throw new Error(`HTTP error status: ${response.status} for ${method} ${url}: ${await response.text()}`);
            }
            const data = await this._parseResponse(response);
            return {
                data,
                status: response.status,
                headers: await this._transformHeaders(response.headers),
            };
        }
        catch (error) {
            throw enhanceNetworkError(error, method, url);
        }
        finally {
            this._stopSpinner();
        }
    }
    _startSpinner() {
        this._inflight++;
        if (this._inflight === 1 && getSettings().interactive) {
            this._spinner.start('Loading...');
        }
    }
    _stopSpinner() {
        this._inflight--;
        if (this._inflight === 0 && getSettings().interactive) {
            this._spinner.stop('');
            // Workaround for @clack/core bug: block() sets stdin to raw mode
            // but the unblock function skips restoring it on Windows.
            // Without this, Ctrl+C won't generate SIGINT after the spinner runs.
            if (process.stdin.isTTY && process.stdin.setRawMode) {
                process.stdin.setRawMode(false);
            }
        }
    }
    async _parseResponse(response) {
        const contentType = response.headers.get('content-type');
        const responseText = await response.text();
        // Handle empty responses or non-JSON content
        if (!responseText || responseText.trim() === '') {
            return null;
        }
        else if (contentType && contentType.includes('application/json')) {
            try {
                return JSON.parse(responseText);
            }
            catch (parseError) {
                throw new Error(`Failed to parse JSON response: ${parseError}`);
            }
        }
        else {
            return responseText;
        }
    }
    async _transformHeaders(headers) {
        const result = {};
        headers.forEach((value, key) => {
            result[key] = value;
        });
        return result;
    }
}
