// @microsoft/power-apps-cli/dist/Verbs/AddFlow.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { addFlowAsync as addFlowActionAsync, getPlayerServiceConfig, } from '@microsoft/power-apps-actions';
import { getCliLogger, getSettings } from '../CliSettings.js';
import { createDefaultStringOption } from '../CliUtils.js';
import { HelpStrings } from '../Constants/HelpStrings.js';
import { CliFS } from '../FS/CliFs.js';
import { AddFlowArguments } from './VerbConstants.js';
/**
 * Add a cloud flow to the current Power Apps code app by retrieving its metadata.
 *
 * @param {ArgumentProvider} argumentProvider
 * @returns {Promise<void>}
 */
export async function addFlowAsync(argumentProvider) {
    const settings = getSettings();
    const { fileConfig } = settings;
    createOptionSchema(argumentProvider);
    await argumentProvider.runVerb();
    const cliFs = new CliFS({
        cwd: process.cwd(),
        allowCwdRead: true,
        allowCwdWrite: true,
    });
    const playerServiceConfig = getPlayerServiceConfig();
    const authenticationProvider = argumentProvider.getAuthenticationProvider();
    const flowId = await argumentProvider.getOption('flowId');
    const context = {
        vfs: cliFs,
        authProvider: authenticationProvider,
        region: playerServiceConfig.region,
        environmentName: playerServiceConfig.environmentName,
        actionsParams: {
            flowId,
        },
        localFilePaths: {
            powerConfigPath: fileConfig.powerConfigPath,
            schemaPath: fileConfig.schemaPath,
            codeGenPath: fileConfig.codeGenPath,
        },
        logger: getCliLogger(),
    };
    const result = await addFlowActionAsync(context);
    if (result.success) {
        console.log('\nFlow added successfully.');
    }
    else {
        console.error(`\nFailed to add flow: ${result.error}`);
    }
}
function createOptionSchema(argumentProvider) {
    argumentProvider.setDescription(HelpStrings.Verbs['Add-flow'].Description);
    argumentProvider.addOptions([
        createDefaultStringOption({
            flags: {
                key: AddFlowArguments.FlowId.name,
                alias: AddFlowArguments.FlowId.alias,
            },
            env: AddFlowArguments.FlowId.envVar,
            message: AddFlowArguments.FlowId.promptMessage,
            description: HelpStrings.Verbs['Add-flow'].Options.FlowId,
        }),
    ]);
}
