// @microsoft/power-apps-cli/dist/Verbs/DeleteDataSource.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { deleteDataSourceAsync as deleteDataSourceActionAsync, getPlayerServiceConfig, } from '@microsoft/power-apps-actions';
import { getCliLogger, getSettings } from '../CliSettings.js';
import { createDefaultStringOption, promptConfirm, promptText } from '../CliUtils.js';
import { HelpStrings } from '../Constants/HelpStrings.js';
import { printSuccess, printWarning } from '../Utils/ConsoleOutput.js';
import { DeleteDataSourceArguments, SharedArguments } from './VerbConstants.js';
export async function deleteDataSourceAsync(argumentProvider) {
    const settings = getSettings();
    const { fileConfig, cliFs } = settings;
    createOptionSchema(argumentProvider);
    await argumentProvider.runVerb();
    const force = await argumentProvider.getOption('force', true);
    if (!force && settings.interactive) {
        const confirmed = await promptConfirm({
            message: 'Are you sure you want to delete this data source?',
        });
        if (!confirmed) {
            printWarning('Delete cancelled.');
            return false;
        }
    }
    const playerServiceConfig = getPlayerServiceConfig();
    // Use the already initialized authentication provider from ArgumentProvider
    const authenticationProvider = argumentProvider.getAuthenticationProvider();
    const memo = (fn) => {
        let promise;
        return () => (promise ??= fn());
    };
    const getApiId = memo(() => argumentProvider.getOption('apiId'));
    const getDataSourceName = memo(() => argumentProvider.getOption('dataSourceName'));
    const getSqlStoredProcedure = memo(() => argumentProvider.getOption('sqlStoredProcedure'));
    const config = {
        vfs: cliFs,
        authProvider: authenticationProvider,
        region: playerServiceConfig.region,
        environmentName: playerServiceConfig.environmentName,
        localFilePaths: fileConfig,
        actionsParams: {
            apiId: getApiId,
            dataSourceName: getDataSourceName,
            sqlStoredProcedure: getSqlStoredProcedure,
        },
        logger: getCliLogger(),
    };
    const returnValue = await deleteDataSourceActionAsync(config);
    if (returnValue) {
        const apiId = await getApiId();
        const isSql = apiId.replace('shared_', '').toLowerCase() === 'sql';
        const dataSourceLabel = isSql ? await getSqlStoredProcedure() : await getDataSourceName();
        printSuccess(`${dataSourceLabel} from API ${apiId} deleted successfully.`);
    }
    return returnValue;
}
function createOptionSchema(argumentProvider) {
    argumentProvider.setDescription(HelpStrings.Verbs['Delete-data-source'].Description);
    argumentProvider.addExamples(HelpStrings.Verbs['Delete-data-source'].Examples);
    argumentProvider.addSwitch({
        flags: { key: 'force', alias: 'f' },
        env: '',
        description: HelpStrings.Verbs['Delete-data-source'].Options.Force,
    });
    argumentProvider.addOptions([
        createDefaultStringOption({
            flags: {
                key: DeleteDataSourceArguments.DataSourceName.name,
                alias: DeleteDataSourceArguments.DataSourceName.alias,
            },
            env: DeleteDataSourceArguments.DataSourceName.envVar
                ? DeleteDataSourceArguments.DataSourceName.envVar
                : SharedArguments.ResourceName.envVar,
            message: DeleteDataSourceArguments.DataSourceName.promptMessage,
            description: HelpStrings.Verbs['Delete-data-source'].Options.DataSourceName,
        }),
        createDefaultStringOption({
            flags: {
                key: SharedArguments.ApiId.name,
                alias: SharedArguments.ApiId.alias,
            },
            env: SharedArguments.ApiId.envVar,
            message: SharedArguments.ApiId.promptMessage,
            description: HelpStrings.Verbs['Delete-data-source'].Options.ApiId,
        }),
        {
            flags: {
                key: SharedArguments.SqlStoredProcedure.name,
                alias: SharedArguments.SqlStoredProcedure.alias,
            },
            env: SharedArguments.SqlStoredProcedure.envVar,
            description: HelpStrings.Verbs['Delete-data-source'].Options.SqlStoredProcedure,
            optionalInNonInteractive: true,
            customPrompt: async () => {
                const hasSqlProcedure = await promptConfirm({
                    message: 'Do you have an SQL procedure?',
                });
                if (hasSqlProcedure) {
                    return await promptText({
                        message: SharedArguments.SqlStoredProcedure.promptMessage,
                    });
                }
                return '';
            },
        },
    ]);
}
