// @microsoft/power-apps-cli/dist/Verbs/FindDataverseApi.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { findDataverseApiAsync as findDataverseApiActionAsync, getPlayerServiceConfig, } from '@microsoft/power-apps-actions';
import { getCliLogger, getSettings } from '../CliSettings.js';
import { createDefaultStringOption } from '../CliUtils.js';
import { HelpStrings } from '../Constants/HelpStrings.js';
import { CliFS } from '../FS/CliFs.js';
import { printWarning } from '../Utils/ConsoleOutput.js';
import { FindDataverseApiArguments } from './VerbConstants.js';
const SEPARATOR_WIDTH = 100;
/**
 * Search Dataverse actions and functions in a Dataverse environment.
 *
 * @param {ArgumentProvider} argumentProvider
 * @returns {Promise<void>}
 */
export async function findDataverseApiAsync(argumentProvider) {
    const settings = getSettings();
    const { fileConfig } = settings;
    createOptionSchema(argumentProvider);
    await argumentProvider.runVerb();
    const cliFs = new CliFS({
        cwd: process.cwd(),
        allowCwdRead: true,
        allowCwdWrite: true,
    });
    const playerServiceConfig = getPlayerServiceConfig();
    const authenticationProvider = argumentProvider.getAuthenticationProvider();
    const searchTerm = await argumentProvider.getOption(FindDataverseApiArguments.SearchTerm.name, true);
    if (!searchTerm) {
        printWarning('Warning: No search term specified. Fetching all Dataverse operations — this may take a while.');
    }
    const context = {
        vfs: cliFs,
        authProvider: authenticationProvider,
        region: playerServiceConfig.region,
        environmentName: playerServiceConfig.environmentName,
        actionsParams: {
            searchTerm,
        },
        localFilePaths: {
            powerConfigPath: fileConfig.powerConfigPath,
            schemaPath: fileConfig.schemaPath,
            codeGenPath: fileConfig.codeGenPath,
        },
        logger: getCliLogger(),
    };
    const operations = await findDataverseApiActionAsync(context);
    if (argumentProvider.isJsonMode()) {
        console.log(JSON.stringify(operations, null, 2));
        return;
    }
    console.log('\n' + '='.repeat(SEPARATOR_WIDTH));
    console.log('Dataverse Operations');
    console.log('='.repeat(SEPARATOR_WIDTH));
    if (operations.length === 0) {
        console.log('\nNo operations found.');
        console.log('='.repeat(SEPARATOR_WIDTH));
        return;
    }
    for (const operation of operations) {
        printCsdlOperation(operation);
    }
    console.log('-'.repeat(SEPARATOR_WIDTH));
    console.log(`Total: ${operations.length} operation(s)`);
    console.log('='.repeat(SEPARATOR_WIDTH));
}
function printCsdlOperation(op) {
    console.log(`\n  ${op.name}  (${op.kind})`);
    const params = op.isBound ? op.parameters.slice(1) : op.parameters;
    if (op.isBound && op.parameters[0]) {
        console.log(`  Bound to: ${op.parameters[0].type}`);
    }
    if (params.length > 0) {
        console.log('  Parameters:');
        for (const p of params) {
            const optional = p.nullable ? '?' : '';
            console.log(`    - ${p.name}${optional}: ${p.type}`);
        }
    }
    if (op.returnType) {
        console.log(`  Returns: ${op.returnType.type}`);
    }
}
function createOptionSchema(argumentProvider) {
    argumentProvider.setDescription(HelpStrings.Verbs['Find-dataverse-api'].Description);
    argumentProvider.addExamples(HelpStrings.Verbs['Find-dataverse-api'].Examples);
    argumentProvider.addOptions([
        createDefaultStringOption({
            flags: {
                key: FindDataverseApiArguments.SearchTerm.name,
                alias: FindDataverseApiArguments.SearchTerm.alias,
            },
            env: FindDataverseApiArguments.SearchTerm.envVar,
            message: FindDataverseApiArguments.SearchTerm.promptMessage,
            description: HelpStrings.Verbs['Find-dataverse-api'].Options.SearchTerm,
        }),
    ]);
}
