// @microsoft/power-apps-cli/dist/Verbs/ListSqlStoredProcedures.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { getPlayerServiceConfig, listSqlStoredProceduresAsync as listSqlStoredProceduresActionAsync, } from '@microsoft/power-apps-actions';
import { getCliLogger, getSettings } from '../CliSettings.js';
import { createDefaultStringOption } from '../CliUtils.js';
import { HelpStrings } from '../Constants/HelpStrings.js';
import { CliFS } from '../FS/CliFs.js';
import { printTable } from '../Utils/ConsoleOutput.js';
import { SharedArguments } from './VerbConstants.js';
export async function listSqlStoredProceduresAsync(argumentProvider) {
    const settings = getSettings();
    const { fileConfig } = settings;
    createOptionSchema(argumentProvider);
    await argumentProvider.runVerb();
    const cliFs = new CliFS({
        cwd: process.cwd(),
        allowCwdRead: true,
        allowCwdWrite: true,
    });
    const playerServiceConfig = getPlayerServiceConfig();
    // Use the already initialized authentication provider from ArgumentProvider
    const authenticationProvider = argumentProvider.getAuthenticationProvider();
    const dsConfig = {
        vfs: cliFs,
        authProvider: authenticationProvider,
        region: playerServiceConfig.region,
        environmentName: playerServiceConfig.environmentName,
        actionsParams: {
            connectionId: await argumentProvider.getOption('connectionId'),
            dataset: await argumentProvider.getOption('dataset'),
        },
        localFilePaths: {
            powerConfigPath: fileConfig.powerConfigPath,
            schemaPath: fileConfig.schemaPath,
            codeGenPath: fileConfig.codeGenPath,
        },
        logger: getCliLogger(),
    };
    const storedProcedures = await listSqlStoredProceduresActionAsync(dsConfig);
    if (argumentProvider.isJsonMode()) {
        console.log(JSON.stringify(storedProcedures?.value ?? [], null, 2));
        return;
    }
    if (!storedProcedures?.value) {
        return;
    }
    printTable({
        title: 'Stored Procedures',
        metadata: [
            `Dataset: ${dsConfig.actionsParams.dataset}`,
            `Connection: ${dsConfig.actionsParams.connectionId}`,
        ],
        columns: [{ header: 'Display Name' }, { header: 'Name' }],
        rows: storedProcedures.value.map((sp) => [sp.DisplayName || '<No Display Name>', sp.Name]),
        emptyMessage: 'No stored procedures found.',
        totalLabel: 'stored procedure(s) found',
    });
}
function createOptionSchema(argumentProvider) {
    argumentProvider.setDescription(HelpStrings.Verbs['List-sqlStoredProcedures'].Description);
    argumentProvider.addExamples(HelpStrings.Verbs['List-sqlStoredProcedures'].Examples);
    argumentProvider.addOptions([
        createDefaultStringOption({
            flags: {
                key: SharedArguments.ConnectionId.name,
                alias: SharedArguments.ConnectionId.alias,
            },
            env: SharedArguments.ConnectionId.envVar,
            message: SharedArguments.ConnectionId.promptMessage,
            description: HelpStrings.Verbs['List-sqlStoredProcedures'].Options.ConnectionId,
        }),
        createDefaultStringOption({
            flags: {
                key: SharedArguments.Dataset.name,
                alias: SharedArguments.Dataset.alias,
            },
            env: SharedArguments.Dataset.envVar,
            message: SharedArguments.Dataset.promptMessage,
            description: HelpStrings.Verbs['List-sqlStoredProcedures'].Options.Dataset,
        }),
    ]);
}
