// @microsoft/power-apps-cli/dist/Verbs/ListCodeApps.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { listAppsAsync } from '@microsoft/power-apps-actions';
import { HelpStrings } from '../Constants/HelpStrings.js';
import { printTable } from '../Utils/ConsoleOutput.js';
export async function getCodeAppsAsync(argumentProvider) {
    argumentProvider.setDescription(HelpStrings.Verbs['List-codeapps'].Description);
    argumentProvider.addExamples(HelpStrings.Verbs['List-codeapps'].Examples);
    // No options, so can just run directly
    await argumentProvider.runVerb();
    const codeApps = await listAppsAsync();
    const filteredCodeApps = codeApps
        .filter((app) => app.appType === 'CodeApp')
        .map((app) => ({
        appId: app.name,
        name: app.properties.displayName,
    }));
    if (argumentProvider.isJsonMode()) {
        console.log(JSON.stringify(filteredCodeApps, null, 2));
        return;
    }
    printTable({
        title: 'Code Apps',
        columns: [{ header: 'App ID', maxWidth: 38 }, { header: 'Display Name' }],
        rows: filteredCodeApps.map((app) => [app.appId, app.name || '<No Name>']),
        emptyMessage: 'No code apps found.',
        totalLabel: 'code app(s) found',
    });
}
