// @microsoft/power-apps-cli/dist/Verbs/ListConnectionReferences.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { getPlayerServiceConfig, listConnectionReferences as listConnectionReferencesActions, } from '@microsoft/power-apps-actions';
import { getCliLogger, getSettings } from '../CliSettings.js';
import { createDefaultStringOption } from '../CliUtils.js';
import { HelpStrings } from '../Constants/HelpStrings.js';
import { CliFS } from '../FS/CliFs.js';
import { printKeyValueList } from '../Utils/ConsoleOutput.js';
import { SharedArguments } from './VerbConstants.js';
export async function listConnectionReferences(argumentProvider) {
    const settings = getSettings();
    const { fileConfig } = settings;
    createOptionSchema(argumentProvider);
    await argumentProvider.runVerb();
    const cliFs = new CliFS({
        cwd: process.cwd(),
        allowCwdRead: true,
        allowCwdWrite: true,
    });
    const playerServiceConfig = getPlayerServiceConfig();
    // Use the already initialized authentication provider from ArgumentProvider
    const authenticationProvider = argumentProvider.getAuthenticationProvider();
    const dsConfig = {
        vfs: cliFs,
        authProvider: authenticationProvider,
        region: playerServiceConfig.region,
        environmentName: playerServiceConfig.environmentName,
        actionsParams: {
            envUrl: await argumentProvider.getOption('orgUrl'),
            solutionId: await argumentProvider.getOption('solutionId'),
        },
        localFilePaths: {
            powerConfigPath: fileConfig.powerConfigPath,
            schemaPath: fileConfig.schemaPath,
            codeGenPath: fileConfig.codeGenPath,
        },
        logger: getCliLogger(),
    };
    const connectionReferences = await listConnectionReferencesActions(dsConfig);
    if (argumentProvider.isJsonMode()) {
        console.log(JSON.stringify(connectionReferences, null, 2));
        return;
    }
    printKeyValueList({
        title: 'Connection References:',
        items: connectionReferences,
        fields: [
            { label: 'Display Name', value: (r) => r.msdyn_displayname },
            { label: 'Name', value: (r) => r.msdyn_name ?? '(not set)' },
            { label: 'Description', value: (r) => r.msdyn_description },
        ],
        emptyMessage: 'No connection references found.',
    });
}
function createOptionSchema(argumentProvider) {
    argumentProvider.setDescription(HelpStrings.Verbs['List-connection-references'].Description);
    argumentProvider.addExamples(HelpStrings.Verbs['List-connection-references'].Examples);
    argumentProvider.addOptions([
        createDefaultStringOption({
            flags: {
                key: SharedArguments.SolutionId.name,
                alias: SharedArguments.SolutionId.alias,
            },
            env: SharedArguments.SolutionId.envVar,
            message: SharedArguments.SolutionId.promptMessage,
            description: HelpStrings.Verbs['List-connection-references'].Options.SolutionId,
        }),
        createDefaultStringOption({
            flags: {
                key: SharedArguments.OrgUrl.name,
                alias: SharedArguments.OrgUrl.alias,
            },
            env: SharedArguments.OrgUrl.envVar,
            message: SharedArguments.OrgUrl.promptMessage,
            description: HelpStrings.Verbs['List-connection-references'].Options.OrgUrl,
        }),
    ]);
}
