// @microsoft/power-apps-cli/dist/Verbs/Push.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { pushApp } from '@microsoft/power-apps-actions';
import { DataverseSystemDataService, getEnvironmentByName, getPlayerServiceConfig, } from '@microsoft/power-apps-actions';
import { getCliLogger, getSettings } from '../CliSettings.js';
import { isGuid, promptConfirm, promptText } from '../CliUtils.js';
import { HelpStrings } from '../Constants/HelpStrings.js';
import { printInfo, printSuccess, printWarning } from '../Utils/ConsoleOutput.js';
import { SharedArguments } from './VerbConstants.js';
export async function pushAsync(argumentProvider) {
    const settings = getSettings();
    const { appConfig, fileConfig, cliFs } = settings;
    argumentProvider.setDescription(HelpStrings.Verbs.Push.Description);
    argumentProvider.addExamples(HelpStrings.Verbs.Push.Examples);
    createOptionSchema(argumentProvider, appConfig?.environmentId, appConfig?.appId);
    await argumentProvider.runVerb();
    if (!appConfig) {
        throw new Error('power.config.json is required to push an app');
    }
    if (!appConfig.environmentId) {
        throw new Error('Environment ID is required in power.config.json to push an app');
    }
    if (!appConfig.region) {
        appConfig.region = 'prod';
    }
    const authenticationProvider = argumentProvider.getAuthenticationProvider();
    const pushConfig = {
        vfs: cliFs,
        actionsParams: {
            solutionId: await argumentProvider.getOption('solutionId', true),
            appName: appConfig.appId,
        },
        environmentName: appConfig.environmentId,
        region: appConfig.region || 'prod',
        authProvider: authenticationProvider,
        localFilePaths: fileConfig,
        logger: getCliLogger(),
        httpClient: getPlayerServiceConfig().httpClient,
    };
    const app = await pushApp(pushConfig);
    printSuccess(`App pushed successfully. You can play your app at ${app.properties.appPlayUri}`);
    printInfo("Hint: Run 'npx power-apps run' to test locally.");
}
function createOptionSchema(argumentProvider, environmentId, appId) {
    argumentProvider.addOption({
        flags: {
            key: SharedArguments.SolutionId.name,
            alias: SharedArguments.SolutionId.alias,
        },
        env: SharedArguments.SolutionId.envVar,
        description: HelpStrings.Verbs.Push.Options.SolutionName,
        optionalInNonInteractive: true,
        customPrompt: async () => {
            if (!environmentId) {
                throw new Error('Environment ID is required');
            }
            if (!appId) {
                const useSpecificSolution = await promptConfirm({
                    message: 'Would you like to specify a solution to push the app into?',
                });
                if (!useSpecificSolution) {
                    const environmentDetails = await getEnvironmentByName(environmentId);
                    const dvService = new DataverseSystemDataService(getPlayerServiceConfig().httpClient);
                    const orgUrl = environmentDetails.properties.linkedEnvironmentMetadata?.instanceUrl;
                    if (!orgUrl) {
                        throw new Error('Organization URL is not available for the specified environment.');
                    }
                    const preferredSolutionId = await dvService.getPreferredSolution(orgUrl);
                    // If we are not able to get a preferred solution, return empty. The app will just not be added to a solution
                    if (!preferredSolutionId) {
                        printWarning('No preferred solution found. App will not be put in a solution. We recommended setting a preferred solution for your environment in the future.');
                        return '';
                    }
                    return preferredSolutionId;
                }
                else {
                    const solutionName = await promptText({
                        message: 'Please enter the solution name or ID:',
                    });
                    if (!solutionName) {
                        return '';
                    }
                    // If it is a GUID, return as is
                    if (isGuid(solutionName)) {
                        return solutionName;
                    }
                    const environmentDetails = await getEnvironmentByName(environmentId);
                    const dvService = new DataverseSystemDataService(getPlayerServiceConfig().httpClient);
                    const orgUrl = environmentDetails.properties.linkedEnvironmentMetadata?.instanceUrl;
                    if (!orgUrl) {
                        throw new Error('Organization URL is not available for the specified environment.');
                    }
                    const solutions = await dvService.getSolutions(orgUrl);
                    const solutionId = solutions.find((solution) => solution.friendlyname.toLowerCase() === solutionName.toLowerCase() ||
                        solution.uniquename.toLowerCase() === solutionName.toLowerCase())?.solutionid;
                    if (!solutionId) {
                        throw new Error(`Solution with name '${solutionName}' not found.`);
                    }
                    return solutionId;
                }
            }
            // If none, return empty
            return '';
        },
    });
}
