// @microsoft/power-apps-cli/dist/Verbs/TelemetrySettings.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import * as fs from 'fs';
import * as path from 'path';
import { CLI_CONFIG_DIRECTORY, USER_SETTINGS_FILE } from '../Constants.js';
import { HelpStrings } from '../Constants/HelpStrings.js';
import { printError, printKeyValueList, printWarning } from '../Utils/ConsoleOutput.js';
/**
 * Modify the user telemetry settings
 * Does not resolve conflicts between options; Just writes whatever is provided
 * @param {ArgumentProvider} argumentProvider
 * @returns {Promise<void>}
 */
export async function modifyTelemetrySettings(argumentProvider) {
    createOptionSchema(argumentProvider);
    await argumentProvider.runVerb();
    // These are switches, so they will be true or undefined
    const enableTelemetry = await argumentProvider.getOption('enable', true);
    const disableTelemetry = await argumentProvider.getOption('disable', true);
    const showSettings = await argumentProvider.getOption('showSettings', true);
    const consoleOnly = await argumentProvider.getOption('consoleOnly', true);
    const outputToConsole = await argumentProvider.getOption('outputToConsole', true);
    const jsonMode = argumentProvider.isJsonMode();
    // if all are undefined, show current settings as default behavior
    if (enableTelemetry === undefined &&
        disableTelemetry === undefined &&
        consoleOnly === undefined &&
        outputToConsole === undefined &&
        showSettings === undefined) {
        displayUserTelemetrySettings(jsonMode);
        return;
    }
    // if both enable and disable are set, throw an error
    if (enableTelemetry && disableTelemetry) {
        throw new Error('Cannot enable and disable telemetry at the same time.');
    }
    if (showSettings) {
        displayUserTelemetrySettings(jsonMode);
    }
    else {
        const telemetryEnabled = enableTelemetry ? true : disableTelemetry ? false : undefined;
        updateUserTelemetrySettings(telemetryEnabled, consoleOnly, outputToConsole);
    }
}
function createOptionSchema(argumentProvider) {
    argumentProvider.setDescription(HelpStrings.Verbs.Telemetry.Description);
    argumentProvider.addExamples(HelpStrings.Verbs.Telemetry.Examples);
    argumentProvider.addSwitches([
        {
            flags: {
                key: 'enable',
                alias: 'te',
            },
            env: 'POWERAPPS_CLI_ENABLE_TELEMETRY',
            description: HelpStrings.Verbs.Telemetry.Options.Enable,
        },
        {
            flags: {
                key: 'disable',
                alias: 'td',
            },
            env: 'POWERAPPS_CLI_DISABLE_TELEMETRY',
            description: HelpStrings.Verbs.Telemetry.Options.Disable,
        },
        {
            flags: {
                key: 'show-settings',
                alias: 'ts',
            },
            env: 'POWERAPPS_CLI_SHOW_SETTINGS',
            description: HelpStrings.Verbs.Telemetry.Options.ShowSettings,
        },
    ]);
    argumentProvider.addOptions([
        {
            flags: {
                key: 'console-only',
                alias: 'tc',
            },
            env: 'POWERAPPS_CLI_CONSOLE_ONLY',
            description: HelpStrings.Verbs.Telemetry.Options.ConsoleOnly,
        },
        {
            flags: {
                key: 'output-to-console',
                alias: 'to',
            },
            env: 'POWERAPPS_CLI_OUTPUT_TO_CONSOLE',
            description: HelpStrings.Verbs.Telemetry.Options.OutputToConsole,
        },
    ]);
}
function displayUserTelemetrySettings(jsonMode) {
    try {
        const userSettingsPath = path.join(CLI_CONFIG_DIRECTORY, USER_SETTINGS_FILE);
        if (fs.existsSync(userSettingsPath)) {
            const fileContent = fs.readFileSync(userSettingsPath, 'utf-8');
            const existingSettings = JSON.parse(fileContent);
            if (jsonMode) {
                console.log(JSON.stringify(existingSettings, null, 2));
                return;
            }
            printKeyValueList({
                title: 'Telemetry Settings:',
                items: [existingSettings],
                fields: [
                    { label: 'Enabled', value: (s) => String(s.enabled) },
                    { label: 'Console Only', value: (s) => String(s.consoleOnly) },
                    { label: 'Output To Console', value: (s) => String(s.outputToConsole) },
                ],
            });
        }
        else {
            if (jsonMode) {
                console.log(JSON.stringify(null));
                return;
            }
            printWarning('No telemetry settings found. Using default settings.');
        }
    }
    catch (error) {
        printError(`Failed to display telemetry settings: ${error.message}`);
    }
}
function updateUserTelemetrySettings(telemetryEnabled, consoleOnly, outputToConsole) {
    // CLI_CONFIG_DIRECTORY, USER_SETTINGS_FILE
    try {
        if (fs.existsSync(CLI_CONFIG_DIRECTORY) === false) {
            fs.mkdirSync(CLI_CONFIG_DIRECTORY, { recursive: true });
        }
        const userSettingsPath = path.join(CLI_CONFIG_DIRECTORY, USER_SETTINGS_FILE);
        let userSettings;
        if (fs.existsSync(userSettingsPath)) {
            const fileContent = fs.readFileSync(userSettingsPath, 'utf-8');
            const existingSettings = JSON.parse(fileContent);
            userSettings = {
                enabled: telemetryEnabled ?? existingSettings.enabled,
                consoleOnly: consoleOnly !== undefined ? consoleOnly === 'true' : existingSettings.consoleOnly,
                outputToConsole: outputToConsole !== undefined
                    ? outputToConsole === 'true'
                    : existingSettings.outputToConsole,
            };
        }
        else {
            userSettings = {
                enabled: telemetryEnabled ?? true,
                consoleOnly: consoleOnly !== undefined ? consoleOnly === 'true' : false,
                outputToConsole: outputToConsole !== undefined ? outputToConsole === 'true' : false,
            };
        }
        fs.writeFileSync(userSettingsPath, JSON.stringify(userSettings, null, 2), 'utf-8');
    }
    catch (error) {
        throw new Error(`Failed to update telemetry settings: ${error.message}`);
    }
}
