// @microsoft/power-apps-cli/dist/Verbs/VerbConstants.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
/**
 * Argument definitions for CLI verbs.
 */
export const SharedArguments = {
    Help: {
        name: 'help',
        alias: 'h',
    },
    ApiId: {
        name: 'api-id',
        alias: 'a',
        promptMessage: 'Please provide the API ID:',
        envVar: 'API_ID',
    },
    SolutionId: {
        name: 'solution-id',
        alias: 's',
        promptMessage: 'Please provide the Solution ID:',
        envVar: 'SOLUTION_ID',
    },
    OrgUrl: {
        name: 'org-url',
        alias: 'u',
        promptMessage: 'Please provide the organization URL:',
        envVar: 'ENV_URL',
    },
    ConnectionId: {
        name: 'connection-id',
        alias: 'c',
        promptMessage: 'Please provide the Connection ID:',
        envVar: 'CONNECTION_ID',
    },
    Dataset: {
        name: 'dataset',
        alias: 'd',
        promptMessage: 'Please provide the dataset:',
        envVar: 'DATASET',
    },
    SqlStoredProcedure: {
        name: 'sql-stored-procedure',
        alias: 'sp',
        promptMessage: 'Please provide the SQL stored procedure name:',
        envVar: 'SQL_STORED_PROCEDURE',
    },
    ResourceName: {
        name: 'resource-name',
        alias: 't',
        promptMessage: 'Please provide the resource name:',
        envVar: 'TABLE',
    },
    Search: {
        name: 'search',
        alias: 's',
        promptMessage: 'Please provide a search term:',
        envVar: 'SEARCH_TERM',
    },
};
export const AddDataSourceArguments = {
    ConnectionRef: {
        name: 'connection-ref',
        alias: 'cr',
        promptMessage: 'Please provide the Connection Reference:',
        envVar: 'CONNECTION_REF',
    },
};
export const DeleteDataSourceArguments = {
    DataSourceName: {
        name: 'data-source-name',
        alias: 'n',
        promptMessage: 'Please provide the data source or table name:',
        envVar: 'DATA_SOURCE_NAME',
    },
};
export const GlobalArguments = {
    EnvironmentId: {
        name: 'environment-id',
        alias: 'e',
        promptMessage: 'Please provide the environment ID:',
        envVar: 'ENVIRONMENT_ID',
    },
    Cloud: {
        name: 'cloud',
        promptMessage: 'Please select the cloud instance:',
        envVar: 'CLOUD_INSTANCE',
    },
};
export const InitArguments = {
    DisplayName: {
        name: 'display-name',
        alias: 'n',
        promptMessage: 'Please provide the display name for the app:',
        envVar: 'APP_DISPLAY_NAME',
    },
    Description: {
        name: 'description',
        alias: 'd',
        promptMessage: 'Please provide a description for the app:',
        envVar: 'APP_DESCRIPTION',
    },
    BuildPath: {
        name: 'build-path',
        alias: 'b',
        promptMessage: 'Please provide the build path for the app:',
        envVar: 'APP_BUILD_PATH',
        initialValue: './dist',
    },
    FileEntryPoint: {
        name: 'file-entry-point',
        alias: 'f',
        promptMessage: 'Please provide the file entry point for the app:',
        envVar: 'APP_FILE_ENTRY_POINT',
        initialValue: 'index.html',
    },
    AppUrl: {
        name: 'app-url',
        alias: 'a',
        promptMessage: 'Please provide the local app URL:',
        envVar: 'APP_URL',
        initialValue: 'http://localhost:3000',
    },
    LogoPath: {
        name: 'logo-path',
        alias: 'l',
        promptMessage: 'Please provide the logo path for the app:',
        envVar: 'APP_LOGO_PATH',
        initialValue: 'Default',
    },
};
export const RunArguments = {
    Port: {
        name: 'port',
        alias: 'p',
        promptMessage: 'Please provide the port number:',
        envVar: 'PORT',
        initialValue: '8080',
    },
    LocalAppUrl: {
        name: 'local-app-url',
        alias: 'l',
        promptMessage: 'Please provide the local app URL:',
        envVar: 'LOCAL_APP_URL',
        initialValue: 'http://localhost:3000',
    },
};
export const ListFlowsArguments = {
    Search: {
        ...SharedArguments.Search,
        promptMessage: 'Please provide a search term to filter flows by name:',
        envVar: 'FLOW_SEARCH',
    },
};
export const AddFlowArguments = {
    FlowId: {
        name: 'flow-id',
        alias: 'f',
        promptMessage: 'Please provide the flow ID:',
        envVar: 'FLOW_ID',
    },
};
export const RemoveFlowArguments = {
    FlowDataSourceName: {
        name: 'flow-name',
        alias: 'n',
        promptMessage: 'Please provide the flow data source name:',
        envVar: 'FLOW_DATA_SOURCE_NAME',
    },
    FlowId: {
        name: 'flow-id',
        alias: 'f',
        promptMessage: 'Please provide the flow ID:',
        envVar: 'REMOVE_FLOW_ID',
    },
};
export const FindDataverseApiArguments = {
    SearchTerm: {
        ...SharedArguments.Search,
        promptMessage: 'Please provide a search term to filter Dataverse operations by name:',
        envVar: 'DATAVERSE_API_SEARCH_TERM',
    },
};
export const RefreshDataSourceArguments = {
    DataSourceName: {
        name: 'data-source-name',
        alias: 'n',
        promptMessage: 'Please provide the data source name to refresh:',
        envVar: 'DATA_SOURCE_NAME',
    },
};
export const AddDataverseApiArguments = {
    ApiName: {
        name: 'api-name',
        key: 'apiName',
        alias: 'n',
        promptMessage: 'Please provide the Dataverse API name:',
        envVar: 'DATAVERSE_API_NAME',
    },
};
