// @microsoft/power-apps-cli/dist/Authentication/NodeMsalAuthenticationProvider.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { InteractionRequiredAuthError, PublicClientApplication } from '@azure/msal-node';
import open from 'open';
import { getAuthority } from '../CliUtils.js';
import { AUTH_CACHE_DIRECTORY } from '../Constants.js';
import { printError, printInfo } from '../Utils/ConsoleOutput.js';
export class NodeMsalAuthenticationProvider {
    _msalClient;
    _tenantId;
    _region;
    constructor() { }
    async initAsync(region) {
        this._region = region;
        const { DataProtectionScope, PersistenceCachePlugin, PersistenceCreator } = await import('@azure/msal-node-extensions');
        const persistenceConfiguration = {
            cachePath: AUTH_CACHE_DIRECTORY + '/msal_cache.json',
            dataProtectionScope: DataProtectionScope.CurrentUser,
            serviceName: 'power-apps',
            accountName: 'power-apps',
            usePlaintextFileOnLinux: false,
        };
        const persistence = await PersistenceCreator.createPersistence(persistenceConfiguration);
        const authConfig = {
            auth: {
                authority: getAuthority(this._region, this._tenantId),
                clientId: '9cee029c-6210-4654-90bb-17e6e9d36617',
            },
            cache: {
                cachePlugin: new PersistenceCachePlugin(persistence),
            },
        };
        this._msalClient = new PublicClientApplication(authConfig);
    }
    async getAccessTokenForResource(resource) {
        const result = await this._acquireTokenInternal(resource + '/.default');
        this._tenantId = result.tenantId;
        return result.accessToken;
    }
    getUserTenantId() {
        return this._tenantId;
    }
    async _acquireTokenInternal(resource) {
        if (!this._msalClient) {
            throw new Error('Authentication not initialized yet');
        }
        const accounts = await this._msalClient.getTokenCache().getAllAccounts();
        if (accounts.length === 1) {
            const silentRequest = {
                account: accounts[0],
                scopes: [resource],
            };
            try {
                const result = await this._msalClient.acquireTokenSilent(silentRequest);
                return result;
            }
            catch (error) {
                if (error instanceof InteractionRequiredAuthError) {
                    return this._msalClient.acquireTokenInteractive(this._getInteractiveLoginRequest(resource));
                }
                throw error;
            }
        }
        else if (accounts.length > 1) {
            accounts.forEach((account) => {
                printInfo(account.username);
            });
            return Promise.reject('Multiple accounts found. Please use the logout command to remove existing accounts.');
        }
        else {
            return this._msalClient.acquireTokenInteractive(this._getInteractiveLoginRequest(resource));
        }
    }
    _getInteractiveLoginRequest(resource) {
        const openBrowser = async (url) => {
            try {
                await open(url, { wait: false });
            }
            catch (error) {
                printError(`Failed to open browser: ${String(error)}`);
            }
        };
        return {
            scopes: [resource],
            openBrowser,
            successTemplate: 'You can close this window now.',
        };
    }
}
