// @microsoft/power-apps-cli/dist/FS/CliFs.js
import fs from 'fs';
import path from 'path';
export class CliFS {
    cwd;
    constructor(config) {
        // Normalize and validate cwd
        this.cwd = path.resolve(config.cwd);
        if (!fs.existsSync(this.cwd)) {
            throw new Error(`Current working directory does not exist: ${this.cwd}`);
        }
    }
    async rmdir(dirPath) {
        const normalizedPath = this.validateWritePath(dirPath);
        await fs.promises.rm(normalizedPath, { recursive: true });
    }
    /**
     * Normalize and resolve a path within the VFS context
     */
    normalizePath(inputPath) {
        if (path.isAbsolute(inputPath)) {
            return path.resolve(inputPath);
        }
        return path.resolve(this.cwd, inputPath);
    }
    /**
     * Validate path for read access
     */
    validateReadPath(inputPath) {
        const normalizedPath = this.normalizePath(inputPath);
        return normalizedPath;
    }
    /**
     * Validate path for write access
     */
    validateWritePath(inputPath) {
        const normalizedPath = this.normalizePath(inputPath);
        return normalizedPath;
    }
    // VFS Interface Implementation
    getCwd() {
        return this.cwd;
    }
    getAbsolutePath(relativePath) {
        return this.normalizePath(relativePath);
    }
    async readFile(inputPath, encoding = 'utf-8') {
        const normalizedPath = this.validateReadPath(inputPath);
        try {
            return await fs.promises.readFile(normalizedPath, encoding);
        }
        catch (error) {
            if (error instanceof Error) {
                if (error.code === 'ENOENT') {
                    throw new Error(`File not found: ${normalizedPath}`);
                }
                throw new Error(`Failed to read file: ${error.message}`);
            }
            throw new Error('Unknown error occurred while reading file');
        }
    }
    async writeFile(inputPath, content, options = {}) {
        const normalizedPath = this.validateWritePath(inputPath);
        const { createDirectories = true, encoding = 'utf-8' } = options;
        // Create parent directories if needed
        if (createDirectories) {
            const parentDir = path.dirname(normalizedPath);
            if (!fs.existsSync(parentDir)) {
                await fs.promises.mkdir(parentDir, { recursive: true });
            }
        }
        await fs.promises.writeFile(normalizedPath, content, { encoding });
    }
    async appendFile(inputPath, content) {
        const normalizedPath = this.validateWritePath(inputPath);
        await fs.promises.appendFile(normalizedPath, content, 'utf-8');
    }
    async readdir(inputPath) {
        const normalizedPath = this.validateReadPath(inputPath);
        return await fs.promises.readdir(normalizedPath);
    }
    async mkdir(inputPath, options = {}) {
        const normalizedPath = this.validateWritePath(inputPath);
        await fs.promises.mkdir(normalizedPath, options);
    }
    async exists(inputPath) {
        const normalizedPath = this.normalizePath(inputPath);
        // Don't check permissions for existence checks - just normalize the path
        try {
            await fs.promises.stat(normalizedPath);
            return true;
        }
        catch {
            return false;
        }
    }
    async stat(inputPath) {
        const normalizedPath = this.validateReadPath(inputPath);
        const stats = await fs.promises.stat(normalizedPath);
        return {
            isFile: () => stats.isFile(),
            isDirectory: () => stats.isDirectory(),
            size: stats.size,
            mtime: stats.mtime,
            ctime: stats.ctime,
        };
    }
    async isFile(inputPath) {
        const stats = await this.stat(inputPath);
        return stats.isFile();
    }
    async isDirectory(inputPath) {
        const stats = await this.stat(inputPath);
        return stats.isDirectory();
    }
    // Path operation methods (delegate to Node.js path module)
    resolve(...paths) {
        return path.resolve(...paths);
    }
    join(...paths) {
        return path.join(...paths);
    }
    relative(from, to) {
        return path.relative(from, to);
    }
    basename(inputPath) {
        return path.basename(inputPath);
    }
    dirname(inputPath) {
        return path.dirname(inputPath);
    }
    unlink(inputPath) {
        const normalizedPath = this.validateWritePath(inputPath);
        return fs.promises.unlink(normalizedPath);
    }
}
