// @microsoft/power-apps-cli/dist/Constants/HelpStrings.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
export const HelpStrings = {
    Global: {
        ProgramDescription: 'Power Apps CLI for managing Power Apps code apps.',
        Cloud: 'Cloud instance to use (e.g., prod, test, etc.).',
        EnvironmentId: 'Environment ID to connect to.',
    },
    Verbs: {
        'Add-data-source': {
            Description: 'Add a data source to the current Power Apps code app.',
            Options: {
                ApiId: 'API identifier.',
                ConnectionId: 'Connection identifier.',
                ConnectionRef: 'Connection reference name.',
                ResourceName: 'Table or resource name.',
                Dataset: 'Dataset identifier.',
                OrgUrl: 'Organization URL.',
                SqlStoredProcedure: 'SQL stored procedure name.',
                SolutionId: 'Solution identifier.',
            },
            Examples: [
                'npx power-apps add-data-source --api-id dataverse --resource-name contact',
                'npx power-apps add-data-source --api-id shared_sql --connection-id <id> --dataset <db> --resource-name <table>',
            ],
        },
        'Delete-data-source': {
            Description: 'Remove a data source from the Power Apps code app.',
            Options: {
                ApiId: 'API identifier associated with the data source.',
                DataSourceName: 'Data source or table name to remove.',
                SqlStoredProcedure: 'SQL stored procedure name to remove.',
                Force: 'Skip confirmation prompt.',
            },
            Examples: [
                'npx power-apps delete-data-source --api-id dataverse --data-source-name contact',
                'npx power-apps delete-data-source --api-id shared_sql --data-source-name <table>',
                'npx power-apps delete-data-source --api-id dataverse --data-source-name contact --force',
            ],
        },
        'Refresh-data-source': {
            Description: 'Refresh one or all data sources in the current Power Apps code app.',
            Options: {
                DataSourceName: 'Name of the specific data source to refresh. Omit to refresh all.',
            },
            Examples: [
                'npx power-apps refresh-data-source',
                'npx power-apps refresh-data-source --data-source-name contact',
                'npx power-apps refresh-data-source -n contact',
            ],
        },
        'Add-dataverse-api': {
            Description: 'Add a Dataverse API (action or function) to the current Power Apps code app.',
            Options: {
                ApiName: 'Dataverse API name to add.',
            },
            Examples: [
                'npx power-apps add-dataverse-api --api-name msdyn_CreateOrder',
                'npx power-apps add-dataverse-api --api-name WinOpportunity',
            ],
        },
        'Find-dataverse-api': {
            Description: 'Search Dataverse actions and functions in the environment by name.',
            Options: {
                SearchTerm: 'Search term to filter Dataverse operations by name.',
            },
            Examples: [
                'npx power-apps find-dataverse-api --search "account"',
                'npx power-apps find-dataverse-api --search "account" --json',
            ],
        },
        Init: {
            Description: 'Initialize a new Power Apps code app.',
            Options: {
                EnvironmentId: 'Environment ID to connect to.',
                Cloud: 'Cloud instance to use (e.g., prod, test, etc.).',
                DisplayName: 'Display name for the app.',
                Description: 'App description.',
                BuildPath: 'Build output path (default: ./dist).',
                FileEntryPoint: 'Entry point file for the app.',
                AppUrl: 'Local URL where the app is hosted.',
                LogoPath: 'Path to the app logo file.',
            },
            Examples: [
                'npx power-apps init --environment-id <id> --display-name "My App"',
                'npx power-apps init --cloud test --environment-id <id> --display-name "My App" --build-path ./build',
            ],
        },
        'List-codeapps': {
            Description: 'List all code apps in the environment.',
            Examples: ['npx power-apps list-codeapps', 'npx power-apps list-codeapps --json'],
        },
        'List-connection-references': {
            Description: 'List all connection references in the environment.',
            Options: {
                SolutionId: 'Solution identifier to filter by.',
                OrgUrl: 'Organization URL.',
            },
            Examples: [
                'npx power-apps list-connection-references --org-url https://org.crm.dynamics.com',
                'npx power-apps list-connection-references --json',
            ],
        },
        'List-datasets': {
            Description: 'List all datasets for a connection.',
            Options: {
                ApiId: 'API identifier.',
                ConnectionId: 'Connection identifier.',
            },
            Examples: [
                'npx power-apps list-datasets --api-id shared_sql --connection-id <id>',
                'npx power-apps list-datasets --api-id shared_sql --connection-id <id> --json',
            ],
        },
        'List-environment-variables': {
            Description: 'List all environment variables in the environment.',
            Options: {
                OrgUrl: 'Organization URL.',
            },
            Examples: [
                'npx power-apps list-environment-variables',
                'npx power-apps list-environment-variables --json',
            ],
        },
        'List-flows': {
            Description: 'List all solution cloud flows that are invokable from Power Apps.',
            Options: {
                Search: 'Filter flows by name (case-insensitive substring match)',
            },
            Examples: [
                'npx power-apps list-flows',
                'npx power-apps list-flows --search "approval" --json',
            ],
        },
        'Add-flow': {
            Description: 'Add a cloud flow to the current Power Apps code app by flow ID.',
            Options: {
                FlowId: 'The unique identifier (GUID) of the flow to add.',
            },
            Examples: ['npx power-apps add-flow --flow-id 00000000-0000-0000-0000-000000000000'],
        },
        'Remove-flow': {
            Description: 'Remove a cloud flow from the current Power Apps code app.',
            Options: {
                FlowDataSourceName: 'The flow data source name to remove (as it appears in power.config). Provide either this or --flow-id.',
                FlowId: 'The flow resource ID (GUID) to remove — same ID used with add-flow. Provide either this or --flow-name.',
            },
            Examples: [
                'npx power-apps remove-flow --flow-name myFlow',
                'npx power-apps remove-flow --flow-id 00000000-0000-0000-0000-000000000000',
            ],
        },
        'List-sqlStoredProcedures': {
            Description: 'List all SQL stored procedures for a dataset.',
            Options: {
                ConnectionId: 'Connection identifier.',
                Dataset: 'Dataset name.',
            },
            Examples: [
                'npx power-apps list-sqlStoredProcedures --connection-id <id> --dataset <db>',
                'npx power-apps list-sqlStoredProcedures --connection-id <id> --dataset <db> --json',
            ],
        },
        'List-tables': {
            Description: 'List all tables for a dataset.',
            Options: {
                ApiId: 'API identifier.',
                ConnectionId: 'Connection identifier.',
                Dataset: 'Dataset name.',
            },
            Examples: [
                'npx power-apps list-tables --api-id shared_sql --connection-id <id> --dataset <db>',
                'npx power-apps list-tables --api-id shared_sql --connection-id <id> --dataset <db> --json',
            ],
        },
        Logout: {
            Description: 'Log out the current user.',
            Examples: ['npx power-apps logout'],
        },
        Push: {
            Description: 'Push the code app to the Power Apps environment.',
            Options: {
                SolutionName: 'Solution name to add the app to.',
            },
            Examples: ['npx power-apps push', 'npx power-apps push --solution-name "MySolution"'],
        },
        Run: {
            Description: 'Run the code app locally.',
            Options: {
                Port: 'Port number for the local server.',
                LocalAppUrl: 'Local URL where the app is hosted.',
            },
            Examples: [
                'npx power-apps run',
                'npx power-apps run --port 8080 --local-app-url http://localhost:3000',
            ],
        },
        Telemetry: {
            Description: 'Manage telemetry settings.',
            Options: {
                Enable: 'Enable telemetry.',
                Disable: 'Disable telemetry.',
                ShowSettings: 'Show current telemetry settings.',
                ConsoleOnly: 'Output telemetry to console only.',
                OutputToConsole: 'Output telemetry to console in addition to sending it.',
            },
            Examples: [
                'npx power-apps telemetry --show-settings',
                'npx power-apps telemetry --enable',
                'npx power-apps telemetry --show-settings --json',
            ],
        },
    },
};
