// @microsoft/power-apps-cli/dist/Constants/VerbRegistry.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { HelpStrings } from './HelpStrings.js';
/**
 * Central registry of all CLI verbs with their descriptions.
 * This array is used to generate the verb table in --help output.
 *
 * IMPORTANT: When adding a new verb, you must:
 * 1. Add to Verb type in Cli.types.ts
 * 2. Add case to executeVerb() switch in Cli.ts
 * 3. Add entry to this VERB_REGISTRY array
 * 4. Add entry to HelpStrings.Verbs in HelpStrings.ts
 */
export const VERB_HELP_REGISTRY = {
    init: { description: HelpStrings.Verbs.Init.Description },
    push: { description: HelpStrings.Verbs.Push.Description },
    run: { description: HelpStrings.Verbs.Run.Description },
    'add-data-source': { description: HelpStrings.Verbs['Add-data-source'].Description },
    'add-dataverse-api': {
        description: HelpStrings.Verbs['Add-dataverse-api'].Description,
    },
    'find-dataverse-api': {
        description: HelpStrings.Verbs['Find-dataverse-api'].Description,
    },
    'list-codeapps': { description: HelpStrings.Verbs['List-codeapps'].Description },
    'list-datasets': { description: HelpStrings.Verbs['List-datasets'].Description },
    'list-tables': { description: HelpStrings.Verbs['List-tables'].Description },
    'list-sqlStoredProcedures': {
        description: HelpStrings.Verbs['List-sqlStoredProcedures'].Description,
    },
    'delete-data-source': { description: HelpStrings.Verbs['Delete-data-source'].Description },
    'list-environment-variables': {
        description: HelpStrings.Verbs['List-environment-variables'].Description,
    },
    'list-connection-references': {
        description: HelpStrings.Verbs['List-connection-references'].Description,
    },
    'list-flows': { description: HelpStrings.Verbs['List-flows'].Description },
    'add-flow': { description: HelpStrings.Verbs['Add-flow'].Description },
    'remove-flow': { description: HelpStrings.Verbs['Remove-flow'].Description },
    'refresh-data-source': { description: HelpStrings.Verbs['Refresh-data-source'].Description },
    logout: { description: HelpStrings.Verbs.Logout.Description },
    telemetry: { description: HelpStrings.Verbs.Telemetry.Description },
};
